File: leetcode/0140-wordBreak.py
class Solution:
    def wordBreak(self, s, wordDict):
        return self._dfs(s, wordDict, {})
    
    def _dfs(self, s, wordDict, cache):
        if s in cache.keys():
            return cache.get(s)
        res = []
        if len(s) == 0:
            res.append("")
            return res
        for word in wordDict:
            if s.startswith(word):
                sub_results = self._dfs(s[len(word):], wordDict, cache)
                for str in sub_results:
                    combine_str = word + ("" if len(str) == 0 else (" " + str))
                    res.append(combine_str)
        cache[s] = res
        return res

File: leetcode/0140-wordBreak.java
class Solution {
    public List<String> wordBreak(String s, List<String> wordDict) {
        return dfs(s, wordDict, new HashMap<String, List<String>>());
    }
    
    private List<String> dfs(String s, List<String> wordDict, Map<String, List<String>> cache) {
        if (cache.containsKey(s)) 
            return cache.get(s);
        List<String> res = new LinkedList<>();
        if (s.length() == 0) {
            res.add("");
            return res;
        }
        for (String word : wordDict) {
            if (s.startsWith(word)) {
                List<String> subResults = dfs(s.substring(word.length()), wordDict, cache);
                for (String str : subResults) 
                    res.add(word + (str.isEmpty() ? "" : " ") + str);
            }
        }
        cache.put(s, res);
        return res;
    }
}
